File: api/routes/calls.js
// routes/calls.js
const express = require('express');
const router = express.Router();
const Call = require('../models/Call');
const User = require('../models/User');

module.exports = () => {
  // Create a new call record and notify receiver
  router.post('/', async (req, res) => {
    console.log('📞 POST /api/calls hit - body:', req.body);
    const { callerId, receiverId, callType } = req.body;
    if (!callerId || !receiverId) return res.status(400).json({ message: 'callerId and receiverId are required' });

    try {
      const newCall = await Call.create({
        callerId,
        receiverId,
        callType: callType || 'audio',
        status: 'ringing',
        startedAt: new Date(),
        duration: 0,
      });
      console.log('✅ Call record created:', newCall._id);

      const callerProfile = await User.findOne({ clerkId: callerId }).lean();
      const payload = {
        ...newCall.toObject(),
        callerName: callerProfile?.username || 'Unknown Caller',
        callerImg: callerProfile?.profilePic || null,
      };

      const onlineUsers = req.app.get('onlineUsers');
      const ioInstance = req.app.get('io');
      const receiverSocketId = onlineUsers.get(receiverId);

      console.log('ℹ️ onlineUsers size:', onlineUsers?.size, 'receiverSocketId:', receiverSocketId);
      if (receiverSocketId && ioInstance) {
        ioInstance.to(receiverSocketId).emit('incoming-call', payload);
        console.log('📤 Emitted incoming-call to receiver socket');
      } else {
        console.log(`⚠️ Receiver ${receiverId} not online`);
      }

      return res.status(201).json(newCall);
    } catch (err) {
      console.error('❌ Error creating call:', err.stack || err.message);
      return res.status(500).json({ error: err.message });
    }
  });

  // Accept a call: mark accepted, notify caller AND signal caller that receiver is ready
  router.put('/accept', async (req, res) => {
    console.log('📥 PUT /api/calls/accept body:', req.body);
    try {
      const { callId } = req.body;
      if (!callId) return res.status(400).json({ message: 'callId is required' });

      const call = await Call.findById(callId);
      if (!call) return res.status(404).json({ message: 'Call not found' });

      call.status = 'accepted';         // matches your Call schema enum
      call.answeredAt = new Date();
      await call.save();

      const ioInstance = req.app.get('io');
      const onlineUsers = req.app.get('onlineUsers');
      const callerSocketId = onlineUsers.get(call.callerId);

      if (callerSocketId && ioInstance) {
        // 1) Tell the caller the call was accepted (UI update)
        ioInstance.to(callerSocketId).emit('call-accepted', call.toObject());
        // 2) Tell the caller that the receiver is ready to negotiate (so caller can create offer)
        ioInstance.to(callerSocketId).emit('receiver-ready', { callId, from: call.receiverId });
        console.log('📤 Emitted call-accepted and receiver-ready to caller socket');
      } else {
        console.warn('⚠️ accept: caller not online or io missing');
      }

      return res.status(200).json(call);
    } catch (err) {
      console.error('❌ SERVER ERROR: /api/calls/accept:', err.stack || err.message);
      return res.status(500).json({ message: 'Internal Server Error' });
    }
  });

  // End/missed
  router.put('/end', async (req, res) => {
    console.log('📥 PUT /api/calls/end body:', req.body);
    try {
      const { callId, status } = req.body;
      if (!callId || !status) return res.status(400).json({ message: 'callId and status are required' });

      const call = await Call.findById(callId);
      if (!call) return res.status(404).json({ message: 'Call not found' });

      const duration = call.status === 'accepted' && status === 'ended'
        ? Math.floor((Date.now() - new Date(call.answeredAt || call.startedAt).getTime())/1000)
        : 0;

      call.status = status;
      call.duration = duration;
      call.endedAt = new Date();
      await call.save();

      const updated = call.toObject();
      const ioInstance = req.app.get('io');
      const onlineUsers = req.app.get('onlineUsers');

      const callerSocket = onlineUsers.get(updated.callerId);
      const receiverSocket = onlineUsers.get(updated.receiverId);
      console.log('ℹ️ callerSocket, receiverSocket', callerSocket, receiverSocket);

      if (callerSocket) ioInstance.to(callerSocket).emit('call-ended', updated);
      if (receiverSocket) ioInstance.to(receiverSocket).emit('call-ended', updated);

      if (status === 'missed' && callerSocket) {
        ioInstance.to(callerSocket).emit('call-declined', { callId });
      }

      return res.status(200).json(updated);
    } catch (err) {
      console.error('❌ SERVER ERROR /api/calls/end:', err.stack || err.message);
      return res.status(500).json({ message: 'Internal Server Error' });
    }
  });

  // Get calls for user
  router.get('/:userId', async (req, res) => {
    const { userId } = req.params;
    console.log('📥 GET /api/calls/:userId', userId);
    try {
      const calls = await Call.find({
        $or: [{ callerId: userId }, { receiverId: userId }]
      }).sort({ createdAt: -1 }).lean();
      console.log('ℹ️ calls found:', calls?.length);
      return res.json(calls);
    } catch (err) {
      console.error('❌ Error fetching calls:', err);
      return res.status(500).json({ message: 'Server error fetching calls' });
    }
  });

  return router;
};

File: api/routes/calls.py
from fastapi import APIRouter, HTTPException, Request
from bson import ObjectId
from datetime import datetime
from motor.motor_asyncio import AsyncIOMotorClient
import os
import store  # <--- IMPORT SHARED STORE

router = APIRouter()

# --- Database Setup ---
MONGO_URI = os.getenv("MONGO_URI", "mongodb+srv://skyees:raj1234%23A@skyees.oabwff3.mongodb.net/skyees?retryWrites=true&w=majority")
client = AsyncIOMotorClient(MONGO_URI)
db = client["skyees"]
calls_collection = db["calls"]
users_collection = db["users"]

# --- Create Call ---
@router.post("")
async def create_call(body: dict, request: Request):
    # 🚨 LOG TAG TO PROVE FILE IS UPDATED
    print("📞 [FINAL_FIX] POST /api/calls hit") 
    
    callerId = body.get("callerId")
    receiverId = body.get("receiverId")
    callType = body.get("callType", "audio")

    new_call = {
        "callerId": callerId,
        "receiverId": receiverId,
        "callType": callType,
        "status": "ringing",
        "startedAt": datetime.utcnow(),
        "createdAt": datetime.utcnow(),
    }
    result = await calls_collection.insert_one(new_call)
    new_call["_id"] = str(result.inserted_id)

    # Fetch Caller Profile
    caller_profile = await users_collection.find_one({"clerkId": callerId})
    payload = {
        **new_call,
        "callerName": caller_profile.get("username") if caller_profile else "Unknown Caller",
        "callerImg": caller_profile.get("profilePic") if caller_profile else None,
    }
    for k, v in payload.items():
        if isinstance(v, datetime): payload[k] = v.isoformat()

    # 🚨 READ FROM STORE (SHARED MEMORY)
    receiver_socket_id = store.online_users.get(receiverId)
    
    print(f"🔍 [FINAL_FIX] Looking for user {receiverId} in store")
    print(f"🔍 [FINAL_FIX] Store has {len(store.online_users)} users")

    if receiver_socket_id and store.io_instance:
        await store.io_instance.emit("incoming-call", payload, to=receiver_socket_id)
        print(f"✅ [FINAL_FIX] Emitted incoming-call to {receiver_socket_id}")
    else:
        print(f"⚠️ [FINAL_FIX] Receiver {receiverId} not online")

    return new_call

# --- Accept Call ---
@router.put("/accept")
async def accept_call(body: dict):
    callId = body.get("callId")
    call = await calls_collection.find_one({"_id": ObjectId(callId)})
    
    if call:
        await calls_collection.update_one(
            {"_id": ObjectId(callId)}, 
            {"$set": {"status": "accepted", "answeredAt": datetime.utcnow()}}
        )
        
        caller_socket_id = store.online_users.get(call["callerId"])
        
        if caller_socket_id and store.io_instance:
            call["_id"] = str(call["_id"])
            for k, v in call.items():
                if isinstance(v, datetime): call[k] = v.isoformat()

            await store.io_instance.emit("call-accepted", call, to=caller_socket_id)
            await store.io_instance.emit("receiver-ready", {"callId": callId, "from": call["receiverId"]}, to=caller_socket_id)

    return {"status": "ok"}

# --- End Call ---
@router.put("/end")
async def end_call(body: dict):
    print("📥 [FINAL_FIX] PUT /api/calls/end hit")
    callId = body.get("callId")
    status = body.get("status")

    if not callId: return {"error": "callId required"}

    call = await calls_collection.find_one({"_id": ObjectId(callId)})
    if not call:
        raise HTTPException(status_code=404, detail="Call not found")

    await calls_collection.update_one(
        {"_id": ObjectId(callId)},
        {"$set": {"status": status, "endedAt": datetime.utcnow()}}
    )

    # 🚨 READ FROM STORE (SHARED MEMORY)
    caller_socket = store.online_users.get(call["callerId"])
    receiver_socket = store.online_users.get(call["receiverId"])

    print(f"🔍 [FINAL_FIX] End Call - CallerSocket: {caller_socket}")
    print(f"🔍 [FINAL_FIX] End Call - ReceiverSocket: {receiver_socket}")

    call["status"] = status
    call["_id"] = str(call["_id"])
    for k, v in call.items():
        if isinstance(v, datetime): call[k] = v.isoformat()

    if caller_socket and store.io_instance:
        await store.io_instance.emit("call-ended", call, to=caller_socket)
        print("✅ [FINAL_FIX] Sent to Caller")
        
    if receiver_socket and store.io_instance:
        await store.io_instance.emit("call-ended", call, to=receiver_socket)
        print("✅ [FINAL_FIX] Sent to Receiver")

    return call